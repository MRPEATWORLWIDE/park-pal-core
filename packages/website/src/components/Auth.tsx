import { AuthProvider } from '#contexts/AuthContext'
import { useSettingsContext } from '#contexts/SettingsContext'
import { AuthReturnType, ClientCredentials, getSalesChannelToken } from '@commercelayer/js-auth'
import { CommerceLayer } from '@commercelayer/react-components'
import { useRouter } from 'next/router'
import { useEffect, useState } from 'react'

type Auth = {
  accessToken: string
  refreshToken?: string
  expires?: number
  tokenType: string
}

const getClientCredentials = (clientId: string, endpoint: string, market: number): ClientCredentials => ({
  clientId,
  endpoint,
  scope: `market:${market}`
})

const getAuth = (market: number): Auth | null => {
  const storeKey = getStoreKey(market)
  return JSON.parse(localStorage.getItem(storeKey) || 'null')
}

const storeAuth = (market: number, authReturn: Awaited<AuthReturnType>): Auth | null => {
  if (!authReturn) {
    return null
  }

  const storeKey = getStoreKey(market)

  const auth: Auth = {
    tokenType: authReturn.tokenType,
    accessToken: authReturn.accessToken,
    expires: authReturn.expires?.getTime(),
    refreshToken: authReturn.refreshToken
  }

  localStorage.setItem(storeKey, JSON.stringify(auth))

  return auth
}

const getStoreKey = <M extends number>(market: M): `clayer_token-market:${M}` => `clayer_token-market:${market}`

const hasExpired = (time: number | undefined): boolean => time === undefined || time < Date.now()

const isValid = (auth: Auth | null): auth is Auth => !hasExpired(auth?.expires)

export const Auth: React.FC<{}> = ({ children }) => {

  const router = useRouter()
  const settings = useSettingsContext()

  const [market, setMarket] = useState<number | undefined>(settings.locale?.isShoppable ? settings.locale?.country.market : undefined)
  const [auth, setAuth] = useState<Auth | null>(null)

  const clientId = process.env.NEXT_PUBLIC_CL_CLIENT_ID
  const endpoint = process.env.NEXT_PUBLIC_CL_ENDPOINT

  useEffect(function updateMarket() {
    if (settings.locale?.isShoppable && settings.locale?.country.market !== market) {
      setMarket(settings.locale?.country.market)
    }
  }, [settings.locale, market])

  useEffect(function updateAccessToken() {
    let isMounted = true

    if (market === undefined || clientId === undefined || endpoint === undefined) {
      setAuth(null)
      return
    }

    const storedAuth: Auth | null = getAuth(market)
    const authIsValid = isValid(storedAuth)

    if (authIsValid) {
      setAuth(storedAuth)
    } else {
      getSalesChannelToken(getClientCredentials(clientId, endpoint, market))
        .then(authReturn => {
          if (isMounted) {
            setAuth(storeAuth(market, authReturn))
          }
        })
    }

    return () => {
      isMounted = false
    }
  }, [market, router.asPath, clientId, endpoint])

  if (!auth || !endpoint) {
    return <>{children}</>
  }

  const { hostname } = new URL(endpoint)
  const [, organization, domain] = hostname.match(/^(.*).(commercelayer.(co|io))$/) || []

  return (
    <AuthProvider accessToken={auth.accessToken} endpoint={endpoint} organization={organization} domain={domain}>
      <CommerceLayer accessToken={auth.accessToken} endpoint={endpoint}>
        <>{children}</>
      </CommerceLayer>
    </AuthProvider>
  )
}
