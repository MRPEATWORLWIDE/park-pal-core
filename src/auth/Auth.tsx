import { getLocale } from '#i18n/locale'
import { AuthReturnType, ClientCredentials, getSalesChannelToken } from '@commercelayer/js-auth'
import { CommerceLayer } from '@commercelayer/react-components'
import { useRouter } from 'next/router'
import { FC, useEffect, useMemo, useState } from 'react'

type Auth = {
  accessToken: string
  refreshToken?: string
  expires?: number
  tokenType: string
}

const getClientCredentials = (market: number): ClientCredentials => ({
  clientId: process.env.NEXT_PUBLIC_CL_CLIENT_ID,
  endpoint: process.env.NEXT_PUBLIC_CL_ENDPOINT,
  scope: `market:${market}`
})

const getAuth = (market: number): Auth | null => {
  const cookieKey = getCookieKey(market)
  return JSON.parse(localStorage.getItem(cookieKey) || 'null')
}

const storeAuth = (market: number, authReturn: Awaited<AuthReturnType>): Auth | null => {
  if (!authReturn) {
    return null
  }

  const cookieKey = getCookieKey(market)

  const auth: Auth = {
    accessToken: authReturn.accessToken,
    refreshToken: authReturn.refreshToken,
    expires: authReturn.expires?.getTime(),
    tokenType: authReturn.tokenType
  }

  localStorage.setItem(cookieKey, JSON.stringify(auth))

  return auth
}

const getCookieKey = <M extends number>(market: M): `clayer_access_token-market:${M}` => `clayer_access_token-market:${market}`

const hasExpired = (time: number | undefined): boolean => time === undefined || time < Date.now()

const isValid = (auth: Auth | null): auth is Auth => !hasExpired(auth?.expires)

export const Auth: FC = ({ children }) => {

  const router = useRouter()

  const locale = useMemo(() => getLocale(router.query.locale as string), [router])

  const [market, setMarket] = useState<number | undefined>(locale?.country?.market)
  const [auth, setAuth] = useState<Auth | null>(null)

  useEffect(function updateMarket() {
    if (locale?.country?.market !== market) {
      setMarket(locale?.country?.market)
    }
  }, [locale])

  useEffect(function updateAccessToken() {
    if (market === undefined) {
      setAuth(null)
      return
    }

    const storedAuth: Auth | null = getAuth(market)
    const authIsValid = isValid(storedAuth)

    if (authIsValid) {
      setAuth(storedAuth)
    } else {
      getSalesChannelToken( getClientCredentials(market) )
        .then( authReturn => setAuth( storeAuth(market, authReturn) ) )
    }
  }, [market])

  if (!auth) {
    return <>{children}</>
  }

  return (
    <CommerceLayer accessToken={auth.accessToken} endpoint={process.env.NEXT_PUBLIC_CL_ENDPOINT}>
      {children}
    </CommerceLayer>
  )
}
