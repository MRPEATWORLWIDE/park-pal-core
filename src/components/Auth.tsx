import { defaultLocale, getLocale } from '#i18n/locale'
import { AuthReturnType, ClientCredentials, getSalesChannelToken } from '@commercelayer/js-auth'
import { CommerceLayer } from '@commercelayer/react-components'
import { useRouter } from 'next/router'
import { useEffect, useMemo, useState } from 'react'
import { AuthProvider } from '#contexts/AuthContext'

type Auth = {
  accessToken: string
  refreshToken?: string
  expires?: number
  tokenType: string
}

const getClientCredentials = (market: number): ClientCredentials => ({
  clientId: process.env.NEXT_PUBLIC_CL_CLIENT_ID,
  endpoint: process.env.NEXT_PUBLIC_CL_ENDPOINT,
  scope: `market:${market}`
})

const getAuth = (market: number): Auth | null => {
  const storeKey = getStoreKey(market)
  return JSON.parse(localStorage.getItem(storeKey) || 'null')
}

const storeAuth = (market: number, authReturn: Awaited<AuthReturnType>): Auth | null => {
  if (!authReturn) {
    return null
  }

  const storeKey = getStoreKey(market)

  const auth: Auth = {
    tokenType: authReturn.tokenType,
    accessToken: authReturn.accessToken,
    expires: authReturn.expires?.getTime(),
    refreshToken: authReturn.refreshToken
  }

  localStorage.setItem(storeKey, JSON.stringify(auth))

  return auth
}

const getStoreKey = <M extends number>(market: M): `clayer_token-market:${M}` => `clayer_token-market:${market}`

const hasExpired = (time: number | undefined): boolean => time === undefined || time < Date.now()

const isValid = (auth: Auth | null): auth is Auth => !hasExpired(auth?.expires)

export const Auth: React.FC<{}> = ({ children }) => {

  const router = useRouter()

  const locale = useMemo(() => getLocale(router.query.locale || defaultLocale), [router])

  const [market, setMarket] = useState<number | undefined>(locale.isShoppable ? locale.country.market : undefined)
  const [auth, setAuth] = useState<Auth | null>(null)

  const endpoint = process.env.NEXT_PUBLIC_CL_ENDPOINT
  const { hostname } = new URL(endpoint)
  const [, organization, domain] = hostname.match(/^(.*).(commercelayer.(co|io))$/) || []

  useEffect(function updateMarket() {
    if (locale.isShoppable && locale.country.market !== market) {
      setMarket(locale.country.market)
    }
  }, [locale, market])

  useEffect(function updateAccessToken() {
    let isMounted = true

    if (market === undefined) {
      setAuth(null)
      return
    }

    const storedAuth: Auth | null = getAuth(market)
    const authIsValid = isValid(storedAuth)

    if (authIsValid) {
      setAuth(storedAuth)
    } else {
      getSalesChannelToken(getClientCredentials(market))
        .then(authReturn => {
          if (isMounted) {
            setAuth(storeAuth(market, authReturn))
          }
        })
    }

    return () => {
      isMounted = false
    }
  }, [market, router.asPath])

  if (!auth) {
    return <>{children}</>
  }

  return (
    <AuthProvider accessToken={auth.accessToken} endpoint={endpoint} organization={organization} domain={domain}>
      <CommerceLayer accessToken={auth.accessToken} endpoint={endpoint}>
        <>{children}</>
      </CommerceLayer>
    </AuthProvider>
  )
}
