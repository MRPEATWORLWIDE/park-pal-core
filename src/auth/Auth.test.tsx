import { render } from '@testing-library/react'
import { Auth } from './Auth'

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

beforeEach(() => {
  useRouter.mockReset()
})

test('should match the snapshot', () => {
  useRouter.mockImplementation(() => ({
    route: '/',
    pathname: '/',
    query: {},
    asPath: '/'
  }))

  const { container } = render(<Auth>Children</Auth>)
  expect(container).toMatchSnapshot()
})
