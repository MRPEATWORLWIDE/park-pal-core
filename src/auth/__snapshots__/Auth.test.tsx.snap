// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<div>
  <div
    data-testid="accessToken"
  >
    accessToken: 
    
  </div>
  <div
    data-testid="endpoint"
  >
    endpoint: 
    
  </div>
</div>
`;
